### PR TITLE
Feat/firebase authentication view model

### PR DESCRIPTION
**Pull request by davidholzwarth**
# Summary

Added the viewModel for authentication with Firebase and corresponding tests.

# Details

Added two methods for the UI to use:
- `signInWithGoogle`
Uses the current context and a coroutine scope, which get passed as parameters. The method goes on to call the `FirebaseAuthRepository`, which provides the backend for the authentication. The viewModel has no onSuccess parameter or return value, but rather updates the state of "currentUser". This means that any UI will need to constantly check for changes in the state of this stateFlow.

- `signOut`
Clears the `currentUser`, effectively signing out the user. This again has no return value but rather changes the state of the `currentUser`stateFlow.

### CODE REVIEW

**Review by rBergCode — CHANGES_REQUESTED** [review]
Hey!
It looks like you have not rebase from main before pushing. Specifically #38 which cleans a lot of the project.
Also, please wait for #37 to be accepted, then merge it to main, then also rebase from main so that there are not the repository changes in this branch.

**TL;DR**: #37 should be merged to main and rebased to this branch before I can review this code.

Thanks!

File: app/src/test/java/ch/hikemate/app/authentication/AuthViewModelTest.kt [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+  private fun setupSignedInUser() {
+    `when`(mockFirebaseAuth.currentUser).thenReturn(mockFirebaseUser)
+    viewModel = AuthViewModel(mockRepository)
+  }
+
+  // Helper functions to setup the ViewModel with a signed-in or signed-out user
+  private fun setupSignedOutUser() {
+    `when`(mockFirebaseAuth.currentUser).thenReturn(null)
+    viewModel = AuthViewModel(mockRepository)
+  }
```
Comment (rBergCode) [inline app/src/test/java/ch/hikemate/app/authentication/AuthViewModelTest.kt:67]: Nice! Love how you did that.

File: app/src/test/java/ch/hikemate/app/authentication/AuthViewModelTest.kt [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+          null
+        }
+        .`when`(mockRepository)
+        .signInWithGoogle(any(), any(), any(), any(), any())
+
+    // Verify that currentUser is initially null
+    assertNull(viewModel.currentUser.first())
+
+    viewModel.signInWithGoogle(this, mockContext)
+
```
Comment (rBergCode) [inline app/src/test/java/ch/hikemate/app/authentication/AuthViewModelTest.kt:114]: I think you forgot to check if the function calls the repository here. You did it in the sign out test.
I suppose you should add something like `verify(mockRepository).signInWithGoogle(any(),any(),any(),any(),any())` or something like that.

File: app/src/test/java/ch/hikemate/app/authentication/AuthViewModelTest.kt [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
```
Comment (rBergCode) [inline app/src/test/java/ch/hikemate/app/authentication/AuthViewModelTest.kt:92]: I think you forgot to check if the function calls the repository here. You did it in the sign out test.
I suppose you should add something like `verify(mockRepository).signInWithGoogle(any(),any(),any(),any(),any())` or something like that.

**Review by rBergCode — APPROVED** [review]
Nice job!
